sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala

```diff
@@ -223,8 +223,9 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
           }
         } else {
           val lit = InternalRow(expected, expected)
+          val dtAsNullable = expression.dataType.asNullable
```

Comment:
why do we need `asNullable` here?

Reply:
Mmh, but a test using a non-nullable expression returning `null` should fail IMHO. So if without this there are test failures, I think those tests are wrong and/or we are hiding potential bugs in the codebase, right? Are there many test failures without this change?

Reply:
Without this change (`asNullable`), `NullPointerException` occurs if `expected` has `null` for a primitive type (e.g. `int`).
If you run `"MapZipWith"` test suite without this PR, you can see `NullPointerException`. If you run `"MapZipWith"` with [the PR](https://github.com/apache/spark/pull/22126), you can see the failure (i.e. hidden potential bug).

Reply:
Without #22126, `MapZipWith` made incorrect `DataType`. However, no test case can detect the failure. This PR can detect that incorrect `DataType` in `MapZipWith` without #22126.

I would appreciate it if you would run the test case without #22126, without and with this PR.

Reply:
Ok, thanks now I think I got it. But I think that the NullPointerException is right to be thrown. That is what a user would get in such a situation. So I don't really think this change should be added.

Reply:
Without this change (`dataType.asNullable`), `MapZipWith` test case would not work correctly. `NullPointerException` occurs when `UnsafeArrayData` for `expected` is created.

Reply:
Mmmh... I am not sure. I just applied you patch on the current master and removed this `asNullable`. I run `MapZipWith` test and it passed.

Reply:
Sorry, I think I have not explained properly my idea/suggestion and I think there is somthing wrong in what you are saying. Without #22126 and with this PR we have two scenarios:

 - Using this `asNullable`: the test fails with incorrect evaluation.
 - Without this `asNullable` the test fails with NullPointerException (it is not passing unexpectedly).

So, IMHO, the right behavior is the second one. For 2 reasons: this is what a user would get in such a scenario; in the first case what you return is a wrong result, which may also be equal to the right one in very weird conditions (with the risk of non-detecting error situation).

So I am strongly against introducing this `asNullable` here. I think we just need to leave it as before.

Reply:
> Without this asNullable the test fails with NullPointerException (it is not passing unexpectedly).

As I wrote [here](https://github.com/apache/spark/pull/20637#discussion_r212805339) as `The unexpected test pass can be recreated without #22126 and with #20367 by removing asNullable.`, the following behavior occurs without #22126,

* Using this `asNullable`: this test fails since the result of `MapZipWith` is incorrect
* Without this `asNullable` the test passes while the result of `MapZipWith` is incorrect

The second one is `not` right behavior. Without `asNullable`, we will have the risk of non-detecting error situation. Therefore, I introduced `asNullable`.

Reply:
I understood what you said but this is not what I have experienced. I applied your patch on current master. Then I reverted #22126. What I saw (and this is what I'd expect indeed) is that:

 - with your patch as it is (ie. with `asNullable`) the test failed because the result is incorrect;
 - without `asNullable` the test failed with a NPE.

Probably someone else can also try and check who is wrong here. cc @ueshin @cloud-fan

Reply:
I think we need `asNullable` there.
From the example of `MapZipWith` with wrong nullabilities (before #22126), the test [HigherOrderFunctionsSuite.scala#L312-L314](https://github.com/apache/spark/blob/d7ef82c53c10e6953d077801045ba1438c6670ab/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HigherOrderFunctionsSuite.scala#L312-L314) should apparently fail and with `asNullable` can detect it, whereas without `asNullable`, the test passes mistakenly.

I can understand that we should fail with a NPE, but as for primitive types with wrong nullabilities (in case where it should be `true` but was `false` by mistake), we use the default value for the type, and create a wrong result instead of throwing a NPE. If without `asNullable`, the expected value will also be converted to the wrong value and the comparison will success between both wrong values, but if with `asNullable`, the expected value will stay the value as is without converting to the wrong value, then we can detect the wrong actual value.

Reply:
thanks for your comment @ueshin. If I am the only one with a different opinion that's ok, but I don't really agree, as it means that we are testing something different from what is run by users. Moreover the same can happen everywhere else in the codebase where we don't check nullabilities.

Anyway, if I am the only one with a different opinion, that's fine. Thanks.